Custom reporter integration

abdulhannanali opened the pull request:
<!-- Thanks for submitting a pull request! Please provide enough information so that others can review your pull request. The two fields below are mandatory. -->

**Summary**
This PR adds the ability to specify custom reporter to be used with Jest. Right now, there's no way for the user to specify the custom reporter they want to use, without tinkering the internals. This let's them specify the Custom reporter(s) if any users want to use with their tests.

This PR relies heavily on the work @dmitriiabramov did in #2115 for adding XUnitReporter. Right now, only the interface through which we can use different reporters has been developed. Reporters such as Nyan and XUnitReporter are work in progress. 

<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->
This PR reuses 

**Test plan**

<!-- Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots / videos if the pull request changes UI. -->
 Integration tests for custom reporters are work in progress.

abdulhannanali commented on packages/jest-config/src/validConfig.js:
```diff
@@ -57,6 +57,9 @@ module.exports = ({
   noStackTrace: false,
   notify: false,
   preset: 'react-native',
+  reporters: [
```
@thymikee Would appreciate some feedback from you on this. There was `ValidationError` being thrown but after a certain number of tries when I added a sample reporters file here, it stopped throwing the error. Shouldn't adding it to `types/Config.js` be suffice

thymikee replied:
nope, `types/Config.js` holds types for Flow (which is stripped with babel for runtime). `jest-validate` checks for actual valid JS object, which in our case is `validConfig.js`.

abdulhannanali replied:
Cool makes sense thanks

abdulhannanali commented on packages/jest-config/src/normalize.js:
```diff
@@ -354,6 +354,9 @@ function normalize(config: InitialConfig, argv: Object = {}) {
       case 'preset':
       case 'replname':
+      case 'reporters':
+        value = _replaceRootDirTags(config.rootDir, config[key]);
```
The previous PR #2115 adds a validation here in order to check the type of the options passed. But after `jest-validate` that doesn't seem to be necessary.

thymikee replied:
It is necessary, `jest-validate` only checks for your config to be valid against an example and throws otherwise, it doesn't modify anything.

abdulhannanali replied:
@thymikee So, If you take a look at the #2115 it adds in a type validation here, which is doing `Array.isArray` to type check if it's an array, but do we really need that here. As I am getting a ValidationError just based on what's there in validConfig.js. 

Here's the ValidationError, I think this error should be suffice. Do we need a custom error here.

```js
● Validation Error:

  Option "reporters" must be of type:
    array
  but instead received:
    string

  Example:
  {
    "reporters": [["./here-it-goes.js", {"option1": true}]]
  }

  Configuration Documentation:
  https://facebook.github.io/jest/docs/configuration.html


```

abdulhannanali commented:
The first custom reporter is going to be a Nyan Cat, as it's trivial to implement

thymikee commented on packages/jest-cli/src/TestRunner.js:
```diff
@@ -396,6 +400,57 @@ class TestRunner {
+            Unexpected custom reporter configuration.
+            Expected to get either a path string or an array of [path, confg]
+            got: ${JSON.stringify(entry)}
+          `);
```
This will be quite heavily indented

abdulhannanali replied:
Yes I can see this dirsrupting the output badly. Instead of using a Custom Error message, Are there any functions in `jest-validate` specifically for this purpose?

dmitriiabramov replied:
i actually always use that kind of error messages and i don't mind them being so heavily indented.
maybe we can add a shared utility function that trims indentation to the minimum value of the block?
probably shouldn't be addressed in this PR tho

abdulhannanali replied:
Yes we can have a function similar to what we are doing with `createConfigError` maybe. I'll open up another PR for this purpose, after some time.

abdulhannanali replied:
@thymikee @dmitriiabramov  Ideally will this be added to `jest-validate` or I guess Validate can be used here, since it's kind of a validation error.

thymikee replied:
It would be great if we could extend validation possibilities. Right now it's indeed pretty limited.

cpojer replied:
There is a module on npm for this purpose called `dedent`. The code style Jest uses here is to separate multiline strings by using string concatenation. Please follow this style here.

abdulhannanali replied:
@cpojer Will do that however, it's going to be a temporary fix, as we need this to be in jest-validate ideally.

thymikee commented on packages/jest-cli/src/TestRunner.js:
```diff
@@ -379,6 +379,10 @@ class TestRunner {
     const config = this._config;
 
+    if (config.reporters) {
+      return this._addCustomReporters(config.reporters);
```
So when you setup any custom reporter, you're getting rid of all default ones, like `DefaultReporter`, `CoverageReporter`, `SummaryReporter`? I'm not happy with that

abdulhannanali replied:
Ooopsie! Let me change this. It was added there for testing purposes. Thanks for pointing it out.

dmitriiabramov replied:
although in some cases that might be a desired behavior. I'm not really sure how to go about that. maybe add another option `noDefaultReporters: true`?

abdulhannanali replied:
@dmitriiabramov Makes sense will add this.

abdulhannanali replied:
@dmitriiabramov The configuration open `noDefaultReporters: true` has been added. One concern here is, Should this option only perform the desired behavior when it's used in tandem with `config.reporters` and otherwise give a warning maybe? Or should it work regardless of config.reporters?

abdulhannanali replied:
I have added the first iteration of this which used regardless of reporters option specified or not.

abdulhannanali commented:
@thymikee I am interested in knowing how we can add a better validation check for the Reporters being imported, in order to check if they have all the given functions we are going to dispatch implemented, maybe we can check if they all extend from the BaseReporter class.

abdulhannanali commented:
@dmitriiabramov  In a Test Suite, Can we consider the the fullPath of the test as the value for the `classname` and `package` attribute within the test.?

abdulhannanali commented:
Not adding it in this PR as @dmitriiabramov suggested, but here's the initial version of XUnitReporter written to be used a Custom Reporter. Any feedback is welcomed. https://gist.github.com/abdulhannanali/6560aa5f53e65ca3cf92fb9ad0836d02

abdulhannanali commented on packages/jest-cli/src/__tests__/TestRunner-test.js:
```diff
@@ -38,6 +38,100 @@ test('.addReporter() .removeReporter()', () => {
+// Checking for the method _addCustomReporters
+// Custom reporters used here are the reporters within the package
+// No extra reporters are included to be used
+describe('_addCustomReporters', () => {
```
@thymikee Hey! I am new to testing. Wrote some unit tests to check the integration. Would you mind giving it a review?

abdulhannanali commented on packages/jest-cli/src/__tests__/TestRunner-test.js:
```diff
@@ -38,6 +38,100 @@ test('.addReporter() .removeReporter()', () => {
+      runner._addCustomReporters(reporters);    
+    };
+
+    expect(addInvalidReporters).toThrow();
```
Calling this passes the test, but at the same time throws an error too, because of the require calls that are done. Is there any way I can avoid this?

dmitriiabramov replied:
you can create a mock file and require it

abdulhannanali replied:
@dmitriiabramov But then the error which is currently thrown won't be thrown. We need to check if the error is thrown. I am doubtful at this point, if really need to test for this, as here we are checking if the require is going to throw invalid error.

abdulhannanali replied:
@dmitriiabramov Sorry! I think this test is way too ambiguous these aren't invalid but non existent reporters. Should be more clearer. Removing this test for now.

abdulhannanali replied:
However, I still would like to clarify, if Mock file is going to be helpful in testing for the case where we are checking if test exists or not.

abdulhannanali commented on packages/jest-cli/src/__tests__/TestRunner-test.js:
```diff
@@ -38,6 +38,115 @@ test('.addReporter() .removeReporter()', () => {
   expect(runner._dispatcher._reporters).not.toContain(reporter);
 });
 
+describe('noDefaultReporters option', () => {
```
Adding test for noDefaltReporters option. As I am using this same option for testing the reporters in the next describe function block too.

abdulhannanali commented on packages/jest-cli/src/__tests__/TestRunner-test.js:
```diff
@@ -38,6 +38,115 @@ test('.addReporter() .removeReporter()', () => {
+  let runner;
+
+  beforeEach(() => {
+    runner = new TestRunner({}, {noDefaultReporters: true}, {});
```
Using noDefualtReporters here to make the addition of tests easier.

abdulhannanali commented:
One of the major concern I need to clear out is how the path resolution for the reporters will work.

abdulhannanali commented:
@cpojer These are the rules we are using 

```json
{
    "rules": {
        "react/react-in-jsx-scope": 0,
        "react/jsx-no-undef": 0,
        "jest/no-focused-tests": 0,
        "no-undef": 0,
        "no-unused-vars": 0,
        "consistent-return": 0
    },
    "description": ".eslintrc-docs.json overrides the rules specified in .eslintrc, to make it more suitable for running on code examples in docs/ folder"
}
```

abdulhannanali commented:
@dmitriiabramov The Reporters in the Jest extend from a BaseReporter, but in Mocha the biggest difference is that the BaseReporter is responsible for keeping track of the whole progress of the tests within it's stats object while the BaseReporter within the Jest is just exposing two more methods and removing the `preRunMessage`.

We can even remove the BaseReporter. However, it has a few advantages, as the code in the TestRunner would break if the reporters don't implement all of the methods on the base reporter. Also, we can additionally check in the code if the reports are inheriting from BaseReporter making things easier for us.

cpojer commented on packages/jest-cli/src/TestRunner.js:
```diff
@@ -364,10 +364,14 @@ class TestRunner {
-    this.addReporter(
-      config.verbose ? new VerboseReporter(config) : new DefaultReporter(),
-    );
+    
```
please don't add more than two newlines at a time :)

bookman25 replied:
Maybe worth adding? http://eslint.org/docs/rules/no-multiple-empty-lines

abdulhannanali replied:
:+1: for `no-multiple-empty-lines` like in StandardJS..

cpojer replied:
@abdulhannanali mind sending a separate PR that adds that lint rule? :)

abdulhannanali replied:
@cpojer Definitely thanks a lot! :+1:

cpojer commented on .vscode/settings.json:
```diff
@@ -2,5 +2,7 @@
-  ]
+],
+"eslint.enable": true,
+"javascript.validate.enable": false
```
hmmm.. should we really add this here? Shall we do it in a separate PR?

abdulhannanali replied:
This is a big mistake. I get this committed with each PR I do, as this is part of my development workflow. Sorry! Will remove it.

cpojer commented on packages/jest-cli/src/TestRunner.js:
```diff
@@ -364,10 +364,14 @@ class TestRunner {
+    
+    // Default Reporters are setup when
+    // noDefaultReporters is false, which is false by default
+    // and can be set to true in configuration.
```
I think this comment is superfluous as it describes what the if statement does. Mind removing it?

abdulhannanali replied:
Yes, need to get over this habit.

abdulhannanali replied:
Comments that describe what statements does are superfluous. Noted! Will remove them!

cpojer commented on packages/jest-cli/src/TestRunner.js:
```diff
@@ -376,12 +380,74 @@ class TestRunner {
+            ${JSON.stringify(reporterPath)}
+          Config:
+            ${JSON.stringify(reporterConfig)}
+        `);
```
Can we print this similarly to how we are already printing warnings/errors, with colors etc? I don't think you need JSON.stringify for the reporterPath, for example.

cpojer commented on packages/jest-config/src/validConfig.js:
```diff
@@ -55,9 +55,13 @@ module.exports = ({
   modulePathIgnorePatterns: ['<rootDir>/build/'],
   modulePaths: ['/shared/vendor/modules'],
   name: 'string',
+  noDefaultReporters: false,
```
I'm unsure about this configuration option. Alternatively we could use

```
reporters: ['summary', 'default'],
```

which gets overwritten when `reporters` is specified in a user configuration. However, I also understand that the ergonomics of that aren't great. Instead of a boolean, I'd almost do something like this:

* `reporters` – this overwrites the default in Jest.
* `additionalReporters` – this just concatenates to the default or the specified `reporters` config.

Not sure if this is necessarily better and maybe my problem is just with the name of this boolean config, though.

cc @dmitriiabramov What do you think?

dmitriiabramov replied:
i actually think it's a really good idea.
so it will be
```js
"reporters": [
  "default",
  ["custom-jest-reporter-package", {configValue: true}]
}
```

```
"additionalReporters": [
  ["custom-jest-reporters-package", {configValue: true}]
]
```
?

cpojer replied:
Ah, if the name of the default reporters is simply "default" (instead of "summary, default") then honestly I think we are fine with "reporters" just overwriting the default:

```js
// Includes the default reporters:
reporters: ["default", "custom-reporter"],
// or excludes them:
reporters: ["custom-reporter"],
// or includes some of them:
reporters: ["summary", "custom-reporter"],
```

This basically would make the default reporter an aggregate of other reporters (jest, summary). We can also change this in the code so that the DefaultReporter wraps the jest (new name for current default reporter?) and summary reporter.

abdulhannanali replied:
@cpojer I like the style @dmitriiabramov included for adding the default reporters. I can remove `.noDefaultReporters` and implement the reporters option first and additionalReporters pretty much going to be based on that so afterwards, even implementing additionalReporters won't be a big problem.

cpojer replied:
Sounds good.

cpojer replied:
Sounds good. Let's start with Dmitrii's suggestion and only implement "reporters" and then see what that looks like and take it from there :)

cpojer commented:
@abdulhannanali super excited you are working on this! Would you mind also including the Nyan reporter example in this diff so that we can see what it may look like? We should make sure that the API that Jest exposes to add reporters is actually useful and won't change too often.

dmitriiabramov commented on packages/jest-cli/src/__tests__/TestRunner-test.js:
```diff
@@ -38,6 +38,115 @@ test('.addReporter() .removeReporter()', () => {
+  test('throws on invalid argument to reporter', () => {
+    expect(() => {
+      runner._addCustomReporters('This should be an array obviously');
+    }).toThrow();
```
you should avoid using `toThrow`, because you can accidentally make a typo in your code and it'll throw with `unexpected token` or something, but the test will still be green because it still throws an error

dmitriiabramov replied:
always use a regular expression or a string to match the message

abdulhannanali replied:
Yes, I will add a regular expression, once I can finalize the type of test that's going to be used for this type checking

abdulhannanali replied:
Thanks a lot for clarifying, I can definitely see what a disaster it can be

dmitriiabramov commented on packages/jest-cli/src/__tests__/TestRunner-test.js:
```diff
@@ -38,6 +38,115 @@ test('.addReporter() .removeReporter()', () => {
+// Checking for the method _addCustomReporters
+// Custom reporters used here are the reporters within the package
+// No extra reporters are included to be used
+describe('_addCustomReporters', () => {
```
this method is prefixed with `_` which signifies that it's private. Generally we try to avoid testing private methods (although in rare cases it's necessary).
Instead of writing many unit tests for these modules i'd encourage you to write a few integration tests (you can check `jest/integration_tests/` directory to see the examples.

Jest integration tests run Jest agains an existing example directory and verify that `stdout/stderr` output matches what's expected. They are a little slow and hard to maintain, but they provide much more value and catch more bugs. At the same time you don't have to change anything in your tests when you refactor code, sinc you're not dealing with implementation details.

abdulhannanali replied:
@dmitriiabramov Thanks a lot that works, based on this I have removed additional tests I added for private method `_addCustomReporters` and will now add integration tests for custom_reporters separately.

cpojer commented:
Btw. there is also this proposal: https://github.com/js-reporters/js-reporters – I'm not sure if we should integrate with this.

bookman25 commented:
Can we also add `--reporters` to the cli?  Right now this only allows it to be specified as a config option, but this would be a big win for jest-editor-support package if it could be passed in from the command line.  Should be simple enough to just add to https://github.com/facebook/jest/blob/master/packages/jest-cli/src/cli/args.js#L51 ?

abdulhannanali commented:
@cpojer Thank you for sharing this `js-reporters` proposal seems quite nice of Qunit team. Something similar also came up in my mind after seeing the Mocha's Test Reporter interface, which is based on an EventEmitter, instead of calling methods on a class.

The EventEmitter interface really has it's advantages regarding the flexibility and making the TestDispatcher independent from how the reporters are implemented. Major parts in the TestRunner regarding the dispatching of methods and the existing runners will need to be updated for this to happen.  I still doubt if we can totally implement the proposal as we'll need to change concept around Suites too. However, it seems exciting, looking forward to get more feedback on this..

I am totally up for making implementing an EventEmitter interface though :)

abdulhannanali commented:
@bookman25 IMO, It's too compicated to support this as a CLI option for now, how about first implementing it as a Config option and finalizing how we deal with it.

abdulhannanali commented:
@cpojer js-reporters seems outdated, and popular frameworks such as Karma and Jasmine doesn't seem to be following this. It can be an overkill following this implementaiton. Jasmine's interface is actually quite similar to what we are already doing.

abdulhannanali commented:
It's also very much focused on Testing frameworks that can be run client side. In order to reuse the reporters there.

cpojer commented:
@abdulhannanali sounds good. Let's build our own standard for reporters then ;)

bookman25 commented on packages/jest-cli/src/TestRunner.js:
```diff
@@ -376,12 +385,75 @@ class TestRunner {
+   *    // Format if we don't want to specify any options
+   * ]
+   */
+  _addCustomReporters(reporters: any) {
```
Doesn't look like this method is being called by anything except tests right now?

dmitriiabramov replied:
yeah. this should be called from the constructor (or `setupReporters()` method).
i also think it should probably take the entire `config` as an argument, since it's aware of the configuration and it's better to encapsulate this logic in once place (i mean the logic of where to get the config from and how to validate it)

abdulhannanali replied:
@dmitriiabramov Yes that helps will make this change

abdulhannanali replied:
@bookman25 Thanks for pointing it out, this will be used by the `_setupReporters`, based on my conversation with Christop and Dmitrii, I am making a few changes on how the reporters options are passed in the configuration.

abdulhannanali commented on packages/jest-cli/src/TestRunner.js:
```diff
@@ -376,12 +385,75 @@ class TestRunner {
+        throw new Error(
+          `Failed to set up reporter: \n` +
+          `Reporter Path:\n` +
+          `\t\t${prettyFormat(reporterPath)}` + 
```
Right now I am using this module `pretty-format` to do the formatting. This is temporary and afterwards I will figure out how I can make more changes.

abdulhannanali commented:
@dmitriiabramov The only reason we seem to have a `BaseReporter` is to make sure all the methods exists so we can save us from method not found errors when we call the method on the given reporter, but  as we move towards custom reporters, a more flexible approach might be needed here. 

The `aggregatedResults` provided by the TestRunner itself, is already doing the work of what Mocha does in it's Base.js class, similar to BaseReporter. So, for the standard reporters it might be okay to extend the BaseReporter class, but do we need to extend in the same way for other reporters. 

We can have a `_validateReporter` method which checks if the given methods exist on the reporter, and if doesn't we just throw an error, or maybe we don't do this too, cos  the reporters should be flexible to implement only the methods they want which is the case when we are implementing simple reporters. 

I am trying this `_validateReporter` approach and if it seems fine to you. Interested in hearing your thoughts about this.

abdulhannanali commented:
@cpojer Can I take all these AggregatedResults related method and form them into one `AggregatedResult class` to be used by `TestRunner.js`?

abdulhannanali commented:
@jest-bot will do, thanks for informing.

bookman25 commented on packages/jest-cli/src/TestRunner.js:
```diff
@@ -273,14 +272,37 @@ class TestRunner {
+   * @param {ReporterConfig} reporters Configuration for all the reporters
+   * @returns {boolean}
+   */
+  _addDefaultReporters(reporters?: ReporterConfig): boolean {
```
Maybe rename this to `_shouldAddDefaultReporters`?

abdulhannanali replied:
Thanks @bookman25 that helps a lot.

thymikee commented on packages/jest-cli/src/ReporterDispatcher.js:
```diff
@@ -0,0 +1,142 @@
+   * @param   {Object} reporter reporter to be validated
+   * @returns {boolean} returns true if the reporter is validated
+   */
+  _validateReporter(reporter: Object) {
```
Can you please remove any JSDOC references? Flow is enough :)

abdulhannanali replied:
@thymikee Oh yes, I can't see much JSDOC used within the application anyways. Removing it will make the application better.

abdulhannanali replied:
@thymikee Yes, JSDoc is certainly not being used much. Will remove it.

thymikee commented on packages/jest-cli/src/TestRunner.js:
```diff
@@ -77,7 +76,7 @@ class TestRunner {
   }
 
-  addReporter(reporter: BaseReporter) {
+  addReporter(reporter: Object) {
```
Shouldn't this be ReporterDispatcher instead of Object?

abdulhannanali replied:
@thymikee I think this should be an Object as we are instantiating the reporters before calling addReporter and passing it the instance.

thymikee commented on packages/jest-config/src/reporterValidationErrors.js:
```diff
@@ -0,0 +1,115 @@
+    reporterIndex:number, reporterValue: any
+) : ValidationError {
+    const errorMessage = (
+        `\tReporter at index ${reporterIndex} must be of type:\n` +
```
I don't recall using tabs in other places?

abdulhannanali replied:
@thymikee Confused too about this one, I think I can remove `\t` as you are right, this hasn't been used previously anywhere in the code. I think I can replace them with two spaces for each \t

thymikee commented on packages/jest-jasmine2/src/treeProcessor.js:
```diff
@@ -41,7 +41,7 @@ function treeProcessor(options: Options) {
 
   return queueRunnerFactory({
-    onException: error => tree.onException.call(tree, error),
+    onException: error => tree.onException(tree, error),
```
`tree` is a context of `onException` you should go with:
```js
onException: error => tree.onException(error)
```

thymikee commented on packages/jest-jasmine2/src/treeProcessor.js:
```diff
@@ -59,7 +59,7 @@ function treeProcessor(options: Options) {
         nodeStart(node);
         await queueRunnerFactory({
-          onException: error => node.onException.call(node, error),
+          onException: error => node.onException(node, error),
```
same here

dmitriiabramov commented on integration_tests/__tests__/custom-reporters-test.js:
```diff
@@ -0,0 +1,145 @@
+    const {onRunComplete, onRunStart, onTestResult, onTestStart} = parsedJSON;
+
+    expect(status).toBe(0);
+    expect(stderr.trim()).toBeFalsy();
```
is that supposed to be an empty string?
if yes then i'd prefer it to be explicit `.toBe('')`

dmitriiabramov commented on integration_tests/__tests__/custom-reporters-test.js:
```diff
@@ -0,0 +1,145 @@
+    expect(onTestResult.called).toBeTruthy();
+    expect(onTestStart.called).toBeTruthy();
+
+    expect(parsedJSON).toMatchSnapshot();
```
you can probably remove other expectations on JSON since you're matching to a snapshot here

abdulhannanali replied:
@dmitriiabramov I thought it would be a good idea in order to make like TOTALLY sure, but that's overdone, will remove them

dmitriiabramov commented on integration_tests/__tests__/custom-reporters-test.js:
```diff
@@ -0,0 +1,145 @@
+    expect(stdout).toMatch('Failed Tests: 0');
+    expect(stdout).toMatch('Total Tests: 1');
+
+    expect(stdout).toMatchSnapshot();
```
also make sure there's no timestapms logged in the stdout, otherwise the snapshot will fail on every new run!

abdulhannanali replied:
@dmitriiabramov Yes I am making sure, that's one of the reasons TestReporter gives it's own little generated JSON instead of just stringifying aggregatedResults and matching that as a snapshot

dmitriiabramov commented on integration_tests/custom_reporters/__tests__/add-fail-test.js:
```diff
@@ -0,0 +1,9 @@
+const add = require('../add');
+
+describe('CustomReporters', () => {
+    test('adds fail', () => {
```
we use 2 spaces for indentation. Can you run `yarn prettier` to format the code?

abdulhannanali replied:
@dmitriiabramov I did run it, for some reason `yarn prettier` only runs for the `__tests__` in the root and not for any other tests.  :confused:

abdulhannanali replied:
The pattern for `'integration_tests/*/**/'` isn't in `scripts/prettier.js`. I'll make a PR for this :+1:

abdulhannanali commented on integration_tests/custom_reporters/package.json:
```diff
@@ -0,0 +1,11 @@
+{
+    "jest": {
+        "reporters": [
```
Ahhh! VSCode sometimes gets to 4 by default. Will have to fix this.

abdulhannanali commented on website/core/center.js:
```diff
@@ -9,11 +9,11 @@ const assign = require('object-assign');
   render() {
-    let {style, ...props} = this.props;
-    style = assign({}, style, {textAlign: 'center'});
+    const {style, ...props} = this.props;
```
Doing this to get all the code linted, this is an eslint error as we were previously reassigning to style.

abdulhannanali commented:
@dmitriiabramov @cpojer I have pushed the working version containing the tests. The custom reporters right now are reusing much of the Reporter functionality Jest already has for it's own reporters. The API to implement custom reporters is exactly the same. It's kind of really providing the hook with which you can add your own custom reporter. 

In order for the reporters to confirm to the standard, I implemented a `_validateReporters` method which validates the reporter by checking if various methods are implemented for sure, but since this makes it very unflexible I removed it. 

I would like to know any other suggestions, I have removed the [WIP] prefix from this PR too. 

Thank you for your help everyone :heart:

abdulhannanali commented:
Circle CI build is failed for some weird yarn reason, I have nothing to do with

bookman25 commented:
Does this allow for reporters being passed in via the CLI?  It didn't look like it, but I might have missed it.  For jest-editor-support it would a big benefit to have CLI support for the reports arg.

abdulhannanali commented:
@bookman25 As discussed before, this PR will only add the capability for the `jest-config`. A separate PR should be opened if we want to add the capability to CLI too. However, I haven't given a lot of thought to it, right now, even though Karma and Ava already does it.. Will think about it once this one is merged.

dmitriiabramov commented:
closing in favor of https://github.com/facebook/jest/pull/3349
